src/ini.pest

```diff
@@ -1,13 +1,17 @@
-WHITESPACE = _{ " " }
+ini = { SOI ~ (section | pair | comment | blank_line)* ~ ws? ~ EOI }
+ws = { (" " | "\t")* }
+
+// section
+section = { ws ~ "[" ~ name ~ "]" ~ ws ~ newline? }
 char = { ASCII_ALPHANUMERIC | "." | "_" | "/" }
+newline = { NEWLINE }
```

Comment:
This allows me to actually capture the newline sequence.

Reply:
Defining `ws` explicitly is slowing down my reading speed - did you try defining `WHITESPACE` again, enabling it to be a "catchy" atomic by adding an `@`-sign before the curly braces?

F.e. `WHITESPACE = @{" " | "\t"}`

(Didn't try that one by myself.)

Reply:
That does work to a point, but it doesn't allow me to capture leading and trailing whitespace, as a rule cannot begin or end with a `~`.

Reply:
Leading or trailing to what? I get them caught...

Reply:
whitespace before a key and after a value is not captured with this technique, as demonstrated in my example link in my above comment. If it works for you, please provide an example.

Reply:
```yaml
- ini
  - WHITESPACE: " "
  - pair
    - name: "username"
    - WHITESPACE: " "
    - equals: "="
    - WHITESPACE: " "
    - value: "noha"
    - WHITESPACE: " "
```

for the line

"` username = noha `"

The first `WHITESPACE` (on global level) is the one before key - the last `WHITESPACE` is the one after the value.

Reply:
(using the linked example)

Reply:
The first whitespace token needs to be nested within the pair. Also, we're not seeing the same thing, because this is what I'm seeing:

```yaml
- ini
  - pair
    - name: "username"
    - WHITESPACE: " "
    - equals: "="
    - WHITESPACE: " "
    - value: "noha"
```